### PR TITLE
[libc++] Remove dependence on <ciso646>

### PR DESCRIPTION
**Pull request by michael-kenzel**
C++23 removed `<ciso646>` from the standard library. The header is used in a few places in order to pull in implementation-specific and feature test macros. The new way of doing that is `<version>`, which should be supported by all supported implementations. This change replaces all those uses of `<ciso646>` with `<version>`.

### CODE REVIEW

**Review by philnik777 — APPROVED** [review]

File: libcxx/test/support/container_debug_tests.h
Comment (mordante) [inline libcxx/test/support/container_debug_tests.h:12]: Below we include `test_macros.h` how about moving that here instead.

File: libcxx/test/support/test_macros.h
Comment (mordante) [inline libcxx/test/support/test_macros.h:23]: The libc++ test suite is intended to work with other standard libraries and all Standard versions. This change might break some Standard library in C++03 mode. Can you use something like. (please make sure this example is formatted properly)

```
#ifndef TEST_IMP_INCLUDED_HEADER
#  ifdef __has_include
#    if __has_include("<version>")
#     include <version>
#     define TEST_IMP_INCLUDED_HEADER
#   endif
# endif
#endif
#ifndef TEST_IMP_INCLUDED_HEADER
#  include <ciso646>
#endif
```

This should keep the test suite portable.

**Review by mordante — CHANGES_REQUESTED** [review]
Thanks for working on this!
A few comments to keep the test suite portable.

File: libcxx/test/support/test_macros.h
Comment (philnik777) [reply libcxx/test/support/test_macros.h:23]: TBH I don't think that makes a lot of sense. Yes, this makes the tests technically a tiny bit less portable, but every standard library should backport `<version>` anyways.

File: libcxx/test/support/container_debug_tests.h
Comment (michael-kenzel) [reply libcxx/test/support/container_debug_tests.h:12]: Given that that file itself doesn't actually seem to use any of the stuff `<version>` is supposed to provide, I guess that would make sense, yeah.

File: libcxx/test/support/test_macros.h
Comment (michael-kenzel) [reply libcxx/test/support/test_macros.h:23]: I guess this comes down to the question of whether the test suite really can and should be fully independent of the implementation it's supposed to test. I certainly don't have any insight into where the libc++ test suite stands on that. If the test suite really only tests the library interface defined in the C++ specification, then I do see the point of making it work with any hypothetical implementation. That little added complexity would probably be justified then. I'll be happy to add that to the PR, I'll just wait for a clear decision on whether it should be added.

Comment (michael-kenzel) [reply libcxx/test/support/test_macros.h:23]: Speaking of this: Is this whole
```cpp
#ifdef __has_include
# if __has_include("<__config>")
#   include <__config>
#   define TEST_IMP_INCLUDED_HEADER
# endif
#endif
```part at the top actually necessary? Shouldn't that stuff already be pulled in by `<version>` or `<ciso646>`?

If so, all of this could probably just become
```cpp
#if __has_include("<version>")
#  include <version>
#else
#  include <ciso646>
#endif
```
?

Comment (philnik777) [reply libcxx/test/support/test_macros.h:23]: The tests are supposed to work on any implementation - with the corresponding defect reports applied. Since there was guidance to support `<version>` in all standards modes from the committee, you could consider it to be a DR, which would mean we can use it for all modes.

Comment (mordante) [reply libcxx/test/support/test_macros.h:23]: @philnik777 do you have a link for "Since there was guidance to support <version> in all standards modes from the committee", mainly since I'm curious.

Comment (philnik777) [reply libcxx/test/support/test_macros.h:23]: I wasn't able to find anything explicitly stating that (I might have missed it, since finding anything in the old meeting minutes is just awful), so I might be wrong that there was explicit guidance.

Comment (ldionne) [reply libcxx/test/support/test_macros.h:23]: FWIW I think the current patch is acceptable. The only way to not support `<version>` in C++03 mode is to explicitly add a `#error` in it based on the C++ version, and I'd be really surprised if any implementation did that.

Comment (mordante) [reply libcxx/test/support/test_macros.h:23]: There are other ways to do it; the header is not required to be a file. However I'm not aware a Standard library not using files. I don't feel too strongly about this so we can leave this part of my change request. I still like the other change request addressed.

Comment (michael-kenzel) [reply libcxx/test/support/test_macros.h]: I have now replaced this whole part
```cpp
// Attempt to get STL specific macros like _LIBCPP_VERSION using the most
// minimal header possible. If we're testing libc++, we should use `<__config>`.
// If <__config> isn't available, fall back to <ciso646>.
#ifdef __has_include
# if __has_include("<__config>")
#   include <__config>
#   define TEST_IMP_INCLUDED_HEADER
# endif
#endif
#ifndef TEST_IMP_INCLUDED_HEADER
#include <ciso646>
#endif
```
with just
```cpp
#ifdef __has_include
# if __has_include("<version>")
#   include <version>
# else
#   include <ciso646>
# endif
#else
#  include <ciso646>
#endif
```
It doesn't seem necessary to explicitly check and include `__config.h` since `<version.h>` already does that for libc++, so this seemed like a valid simplification?

File: libcxx/test/support/container_debug_tests.h
Comment (michael-kenzel) [reply libcxx/test/support/container_debug_tests.h]: I've pushed an update that simply removes this include and leave it to `test_macros.h`.

Comment (ldionne) [reply libcxx/test/support/container_debug_tests.h]: We need to include `test_macros.h` before we try to use things like `_LIBCPP_VERSION` then, though.

Comment (michael-kenzel) [reply libcxx/test/support/container_debug_tests.h]: Sorry, very stupid mistake, forgot to build the tests… fixed…

**Review by ldionne — APPROVED** [review]
This version of the patch looks good to me. Thanks for the cleanup.

Perhaps give 1-2 days for @mordante to see this in case he still has feedback, but I think this should be fairly uncontroversial.

File: libcxx/test/support/test_macros.h
Comment (ldionne) [reply libcxx/test/support/test_macros.h]: `<__config>` is only a libc++ thing, and we make no attempt to support another hypothetical library that would define that header. It is diligent from you to consider that case, but yeah that's not really a concern here.

**Comment by michael-kenzel** [comment]
I'm not quite sure what to do about these failed checks. From what I can gleem from the logs, this seems to be some issue compiling some parts of libc++ with gcc. But I'm not sure how my changes would affect that. I already tried rebasing onto the latest main branch, but that doesn't help. And trying to run the tests here locally, I'm running into other problems with gcc compiling libc++, even without my changes.

**Comment by ldionne** [comment]
Oh, just rebase onto `main`, those have been fixed by reverting a patch.

**Comment by michael-kenzel** [comment]
Ok, thanks, that seems to have fixed it now. I guess I was just very unlucky with my first rebase attempt yesterday.

**Review by ldionne — APPROVED** [review]

**Comment by ldionne** [comment]
You have the setting enabled to hide your email on Github, which means that commits are authored as `15786918+michael-kenzel@users.noreply.github.com`. We're trying to move away from that. Please disable the setting so I can merge.

**Review by mordante — APPROVED** [review]
LGTM, thanks for the improvements!

**Comment by michael-kenzel** [comment]
> You have the setting enabled to hide your email on Github, which means that commits are authored as `15786918+michael-kenzel@users.noreply.github.com`. We're trying to move away from that. Please disable the setting so I can merge.

Ah, was not aware of that. Should be turned off now.

**Pull request merged**
